src/ripple/protocol/impl/STXChainAttestationBatch.cpp

```diff
@@ -0,0 +1,487 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/protocol/STXChainAttestationBatch.h>
+
+#include <ripple/basics/Log.h>
+#include <ripple/basics/StringUtilities.h>
+#include <ripple/json/json_get_or_throw.h>
+#include <ripple/protocol/Indexes.h>
+#include <ripple/protocol/PublicKey.h>
+#include <ripple/protocol/SField.h>
+#include <ripple/protocol/STAccount.h>
+#include <ripple/protocol/STAmount.h>
+#include <ripple/protocol/STArray.h>
+#include <ripple/protocol/STObject.h>
+#include <ripple/protocol/Serializer.h>
+#include <ripple/protocol/jss.h>
+
+#include <algorithm>
+
+namespace ripple {
+
+namespace Attestations {
+
+TER
+checkAttestationPublicKey(
+    ReadView const& view,
+    std::unordered_map<AccountID, std::uint32_t> const& signersList,
+    AccountID const& attestationSignerAccount,
+    PublicKey const& pk,
+    beast::Journal j)
+{
+    if (!signersList.contains(attestationSignerAccount))
+    {
+        return tecNO_PERMISSION;
+    }
+
+    AccountID const accountFromPK = calcAccountID(pk);
+
+    if (auto const sleAttestationSigningAccount =
+            view.read(keylet::account(attestationSignerAccount)))
+    {
+        if (accountFromPK == attestationSignerAccount)
+        {
+            // master key
+            if (sleAttestationSigningAccount->getFieldU32(sfFlags) &
+                lsfDisableMaster)
+            {
+                JLOG(j.trace()) << "Attempt to add an attestation with "
+                                   "disabled master key.";
+                return tecXCHAIN_BAD_PUBLIC_KEY_ACCOUNT_PAIR;
```

Comment:
Consider returning `tefMASTER_DISABLED` instead of `tecXCHAIN_BAD_PUBLIC_KEY_ACCOUNT_PAIR`.  That's what we would return in a similar situation with regular multisigning.

Reply:
I think this should be a `tec` code, not a `tef` - we do want to forward the transaction. I suppose I could create a special `tec` codes for each type of "you can't use this key" failure, but one code for "you can't use that key" seems reasonable too.